### ISSUE TITLE
"Cluster agent is not ready error" with newly provisioned nodes, and error never goes away, cluster never starts

### ISSUE THREAD

**Issue by Goldcap** [rancher/rancher#30080]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug: Clusters Never Start, CoreDNS can't resolve cluster hostname, so healthchecks fail. Rancher is working properly, nodes working properly, but for whaterver reason CoreDNS is a fail on lookups.

Note:  After seeing this issue surface here in a few different configurations, and frankly never getting resolution, I'm starting a new ticket as this blocks me from creating BareMetal clusters on CentOS 7.8.

Thanks in advance for looking into this! This probably has a simple resolution (dnsPolicy???) but I wasn't able to figure it out after a few days of trying.

**Steps to reproduce (least amount of steps as possible):**
Rancher installed on a node, v2.5.2, then I create a new cluster, (tried multiple CNI configs), three brand-new never used VSphere VM's on CentOS 7.8 with SELinux enabled. All hosts have both A and CNAMES, and the CNAMES resolve for our FQDN which has a valid certificate.

**Result:**
"Cluster health check failed: cluster agent is not ready"
/etc/resolve.conf has our local DNS
/etc/docker/daemon.js has our local DNS
`dig` returns all records properly on host nodes
the rancher cluster resolves properly using kubectl:
`kubectl run -it --rm --restart=Never busybox --image=busybox:1.28 -- nslookup our.fqdn.com`

Clusters never get past the health check, and can't resolve the primary rancher domain url.

**Other details that may be helpful:**
I tried to modify the CoreDNS config in the cluster YAML:
```
dns:
    upstreamnameservers:
      - dns1ip
      - dns2ip
```
I can perhaps try some additional things, but as the cluster never starts, I can't use kubectl to play around with it.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
2.5.2
- Installation option (single install/HA):
Single install

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):
BareMetal VSphere VMs on CentOS 7.8 with SELinux
- Machine type (cloud/VM/metal) and specifications (CPU/memory):
4 CPU 16 GIG RAM 4 nodes (Rancher + 3 Cluster)
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8+k3s1", GitCommit:"6b595318666804506c19cfc3d3d228423d38fab1", GitTreeState:"clean", BuildDate:"2020-08-14T20:42:29Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
Client: Docker Engine - Community
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 17:03:45 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:02:21 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```

**Comment by superseb** [comment]
Please supply how many nodes with what roles you are adding. Also please supply the same info as requested in https://github.com/rancher/rancher/issues/29715

**Comment by Goldcap** [comment]
Yes, I hope I'm covering what you need with the following, let me know if this isn't the case:

###Background
1. We have an internal DNS, it's running our entire infrastructure, so I'm sure it's working. (500+ Employees use it for 6+ years)
2. We have a FQDN, and a Cert, so I don't think it's our DNS/Host setup, and the Rancher UI works great.
3. All nodes are "wiped" before installation, and we remove the following directories (if they exist, which they won't)
```/var/lib/etcd /etc/kubernetes /etc/cni /opt/cni /var/lib/cni /var/run/calico /opt/rke /var/lib/docker /var/lib/rancher```

I can resolve in the Rancher local node, but not in the cluster containers, despite all of the cluster nodes resolving properly.

###Node Configs
**Can you explain the use-case that you have for just provisioning etcd + controlplane? 
Are you not running workers at all? Can you confirm it works when you add a worker?**
 
I've tried a number of configs:
1. One etc/cp/worker
2. Three etc/cp/worker
3. One etc/cp and one worker
4. Three etc/cp and three workers

In all cases where workers are separate, I provision the etc/cp nodes first, then add worker nodes.

Thank you! Happy to provide any additional details, or do any debugging that might seem to help.

**Comment by superseb** [comment]
Please supply the outputs of the requested commands:
* https://github.com/rancher/rancher/issues/29715#issuecomment-716640246
* https://github.com/rancher/rancher/issues/29715#issuecomment-717141811

**Comment by Goldcap** [comment]
Here's the control plane docker list:
[controlplane.txt](https://github.com/rancher/rancher/files/5612388/controlplane.txt)

And likewise the worker docker list:
[worker.txt](https://github.com/rancher/rancher/files/5612390/worker.txt)

It's 100% DNS resolution, here's the error:
```
INFO: Environment: CATTLE_ADDRESS=10.42.2.2 CATTLE_CA_CHECKSUM=863d7afb904d2aed3b00577e256dc63a2c0d46f5d55ca7e5c08f892582e6ca08 CATTLE_CLUSTER=true CATTLE_FEATURES= CATTLE_INTERNAL_ADDRESS= CATTLE_IS_RKE=true CATTLE_K8S_MANAGED=true CATTLE_NODE_NAME=cattle-cluster-agent-669dc8f987-njjhn CATTLE_SERVER=https://rancher-k8s.mydomain.org
INFO: Using resolv.conf: nameserver 10.43.0.10 search cattle-system.svc.cluster.local svc.cluster.local cluster.local mydomain.org options ndots:5
ERROR: https://rancher-k8s.mydomain.org/ping is not accessible (Could not resolve host: rancher-k8s.mydomain.org)
```

I can't do manual DNS resolution, as the cluster agent keeps dying and restarting, I can't exec in to use dig:
```
224a4cdc7c36        444dfdf0ae3c                           "run.sh"                 About a minute ago   Exited (1) About a minute ago                       k8s_cluster-register_cattle-cluster-agent-669dc8f987-njjhn_cattle-system_d415545d-c8fc-4d48-b779-76899036bf71_7
```

The hosts themselves resolve just fine:
```
;; ANSWER SECTION:
rancher-k8s.mydomain.org. 604800 IN    CNAME   rancher.k8s.mydomain.org.
rancher.k8s.mydomain.org. 604800 IN    A       10.10.10.10
```

Our error from CoreDNS is this:
```
E1129 15:49:12.363010       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.43.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.43.0.1:443: connect: no route to host
```

Again, this works with the prior version of Rancher, (even an upgrade), this only happens when I deploy a new 2.5.2 cluster. So our sad CoreDNS seems to be the culprit, and not sure why we're blocked on that endpoint. All host nodes have 443 open.

Sorry this took so long to do, had to create new VM's, deploy a fresh 2.5.2 cluster.  Thanks for looking into it!

**Comment by superseb** [comment]
Running k8s 1.19 is recommended to disable firewalld, can you confirm this is disabled? 1.19 is also only available in 2.5.x, I guess we can also confirm its not 2.5.x related if you can successfully provision a 1.18 cluster in 2.5.x.

**Comment by Goldcap** [comment]
@superseb Yes I foolishly thought that since `firewalld` worked with 1.18 we'd be ok, apologies. For anyone following, I used this Docker article to configure `iptables` (note there's a typo in the systemd service path):

https://docs.docker.com/network/iptables/

But running the following command we're still getting errors:

```docker ps -a -q --filter "label=io.kubernetes.container.name=cluster-register" | while read container; do echo "=> $container"; docker logs $container; done

INFO: Environment: CATTLE_ADDRESS=10.42.0.3 CATTLE_CA_CHECKSUM=863d7afb904d2aed3b00577e256dc63a2c0d46f5d55ca7e5c08f892582e6ca08 CATTLE_CLUSTER=true CATTLE_FEATURES= CATTLE_INTERNAL_ADDRESS= CATTLE_IS_RKE=true CATTLE_K8S_MANAGED=true CATTLE_NODE_NAME=cattle-cluster-agent-58547fbdbc-lk7c2 CATTLE_SERVER=https://rancher.somedomain.org
INFO: Using resolv.conf: nameserver 10.43.0.10 search cattle-system.svc.cluster.local svc.cluster.local cluster.local somedomain.org options ndots:5
ERROR: https://rancher.somedomain.org/ping is not accessible (Could not resolve host: rancher.somedomain.org)```

We can dig the rancher node from the VM just fine:

```
$ dig rancher.somedomain.org

;; ANSWER SECTION:
rancher.somedomain.org. 604800 IN    CNAME   rancher.somedomain.org.
rancher.root.somedomain.org. 604800 IN    A       10.10.10.10
```

I would note some mTLA issue is happening with our very awesome and not strange DigiCert, which works in the Rancher UI, and is in use in a number of public settings, but from the master node:

```wget https://rancher.somedomain.org/ping
--2020-12-04 18:20:53--  https://rancher.somedomain.org/ping
Resolving rancher.somedomain.org (rancher.somedomain.org)... 10.10.10.10
Connecting to rancher.somedomain.org (rancher.somedomain.org)|10.10.10.10|:443... connected.
ERROR: cannot verify rancher.somedomain.org's certificate, issued by ‘/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=Thawte RSA CA 2018’:
  Unable to locally verify the issuer's authority.
To connect to rancher.somedomain.org insecurely, use `--no-check-certificate'.
```

So in the interest of getting on with it, I'll just deploy using 1.18 and then upgrade, which seems to work fine, even using `firewalld`.

**Comment by superseb** [comment]
The error is the same, I would still need the outputs of the commands listed above to help. Certificate is not an issue (yet) if it can't resolve where to go.

**Comment by Goldcap** [comment]
Hey! I had some additional `iptables` related issues, so on a lark ran `iptables -F` and now it's working :+1: 

I'm gonna blame `firewalld` and call this one closed. For posterity, anyone using CentOS 7.x.x, here's exactly what I did, might be a bit of psuedo-code in here so use at your own risk:

```
echo "*filter
:INPUT ACCEPT [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:FILTERS - [0:0]
:DOCKER-USER - [0:0]

-F INPUT
-F DOCKER-USER
-F FILTERS

-A INPUT -i lo -j ACCEPT
-A INPUT -p icmp --icmp-type any -j ACCEPT
-A INPUT -j FILTERS

-A DOCKER-USER -i ens33 -j FILTERS

-A FILTERS -m state --state ESTABLISHED,RELATED -j ACCEPT
-A FILTERS -m state --state NEW -s 1.2.3.4/32 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 22 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 23 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 8080 -j ACCEPT
-A FILTERS -m state --state NEW -m udp -p udp --dport 500 -j ACCEPT
-A FILTERS -m state --state NEW -m udp -p udp --dport 4500 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 4505:4506 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 5666 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 80 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 443 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 2376 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 2379 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 2380 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 6443 -j ACCEPT
-A FILTERS -m state --state NEW -m udp -p udp --dport 8472 -j ACCEPT
-A FILTERS -m state --state NEW -m udp -p udp --dport 10254 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 9099 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 10250 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 10254 -j ACCEPT
-A FILTERS -m state --state NEW -m tcp -p tcp --dport 30000:32767 -j ACCEPT
-A FILTERS -m state --state NEW -m udp -p udp --dport 30000:32767 -j ACCEPT
-A FILTERS -j REJECT --reject-with icmp-host-prohibited

COMMIT" > /etc/iptables.conf

iptables-restore -n /etc/iptables.conf

echo "[Unit]
Description=Restore iptables firewall rules
Before=network-pre.target

[Service]
Type=oneshot
ExecStart=/sbin/iptables-restore -n /etc/iptables.conf

[Install]
WantedBy=multi-user.target" > /etc/systemd/system/iptables.service

systemctl disable firewalld
systemctl stop firewalld

systemctl enable iptables
systemctl start iptables
iptables -F

systemctl stop docker

rm -rf /var/lib/docker/*
rm -rf /var/lib/rancher/*
cleanupdirs="/var/lib/etcd /etc/kubernetes /etc/cni /opt/cni /var/lib/cni /var/run/calico /opt/rke"
for dir in $cleanupdirs; do
  echo "Removing $dir"
  rm -rf $dir
done

systemctl start docker
```

I then registered etc/cp nodes, and separate worker nodes, after about five minutes we were all set.

Seb, you have the patience of a wizard and thank you!!
